server/src/test/java/org/elasticsearch/cluster/metadata/DataStreamTests.java

```diff
@@ -510,6 +676,338 @@ public void testAddBackingIndexWithAliases() {
+            () -> ds1.addFailureStoreIndex(builder.build(), indexToAdd)
+        );
+        assertThat(
+            e.getMessage(),
+            equalTo(
+                String.format(
+                    Locale.ROOT,
+                    "cannot add index [%s] to data stream [%s] because it is already a backing index on data stream [%s]",
+                    indexToAdd.getName(),
+                    ds1.getName(),
+                    ds2.getName()
+                )
+            )
+        );
+    }
+
+    public void testAddExistingFailureStoreIndex() {
+        int numBackingIndices = randomIntBetween(2, 32);
+        int numFailureIndices = randomIntBetween(2, 32);
+        String dataStreamName = randomAlphaOfLength(10).toLowerCase(Locale.ROOT);
+        final long epochMillis = System.currentTimeMillis();
+
+        List<Index> indices = new ArrayList<>(numBackingIndices);
+        for (int k = 1; k <= numBackingIndices; k++) {
+            indices.add(new Index(DataStream.getDefaultBackingIndexName(dataStreamName, k, epochMillis), UUIDs.randomBase64UUID(random())));
+        }
+        List<Index> failureIndices = new ArrayList<>(numFailureIndices);
+        for (int k = 1; k <= numFailureIndices; k++) {
+            failureIndices.add(
+                new Index(DataStream.getDefaultFailureStoreName(dataStreamName, k, epochMillis), UUIDs.randomBase64UUID(random()))
+            );
+        }
+
+        Metadata.Builder builder = Metadata.builder();
+        for (int k = 1; k <= numBackingIndices; k++) {
+            IndexMetadata im = IndexMetadata.builder(indices.get(k - 1).getName())
+                .settings(settings(IndexVersion.current()))
+                .numberOfShards(1)
+                .numberOfReplicas(1)
+                .build();
+            builder.put(im, false);
+        }
+        for (int k = 1; k <= numFailureIndices; k++) {
+            IndexMetadata im = IndexMetadata.builder(failureIndices.get(k - 1).getName())
+                .settings(settings(IndexVersion.current()))
+                .numberOfShards(1)
+                .numberOfReplicas(1)
+                .build();
+            builder.put(im, false);
+        }
+        DataStream original = DataStreamTestHelper.newInstance(dataStreamName, indices, failureIndices);
+        builder.put(original);
+        Index indexToAdd = randomFrom(failureIndices.toArray(Index.EMPTY_ARRAY));
+
+        DataStream updated = original.addFailureStoreIndex(builder.build(), indexToAdd);
+        assertThat(updated.getName(), equalTo(original.getName()));
+        assertThat(updated.getGeneration(), equalTo(original.getGeneration()));
+        assertThat(updated.getIndices().size(), equalTo(numBackingIndices));
+        assertThat(updated.getFailureIndices().size(), equalTo(numFailureIndices));
+        for (int k = 0; k < numFailureIndices; k++) {
+            assertThat(updated.getFailureIndices().get(k), equalTo(original.getFailureIndices().get(k)));
+        }
```

Comment:
Couldn't we directly do `assertThat(updated, equalTo(original))` here?

Reply:
Yeah, that seems like a bizarre bit of copy paste from a previous test case as a starting point. I'll clean it up.